### PR TITLE
[FVM] use safe randomness for script execution

### PR DESCRIPTION
**Pull request by tarakby**
- update FVM script execution to use a pseudo-random number generator when calling Cadence random function and delete the dummy random number generator
 - update FVM script execution to use the same entropy source as transaction execution (safe random beacon). This is done by adding new `EntropyProviderPerBlock` field to `QueryExecutor`.
 - add new `ScriptInfoParams` field to `EnvironmentParams` struct to provide the executed script code and its hash (ID)
 - use the script ID to diversify the PRNG used for randomness per script. This does not add entropy to randomness but prevents generating the same random sequence for all scripts.
 - improve transaction tests and add new script tests.
 
 Side update:
 - delete `DefaultHasher` structure that can cause a bottleneck in execution because of its locks. Use a default hash function instead.

### CODE REVIEW

File: fvm/script.go [hunk 1]
```diff
@@ -24,7 +24,7 @@ type ScriptProcedure struct {
 }
 
 func Script(code []byte) *ScriptProcedure {
-	scriptHash := hash.DefaultHasher.ComputeHash(code)
+	scriptHash := hash.DefaultComputeHash(code)
```
Comment (janezpodhostnik) [inline fvm/script.go:27]: Not a comment on this PR:

In my mind the ID of the script also contained the script arguments... Its interesting that it doesn't. As a consequence a script like a get balance script that accepts an address as an argument will always have the same ID no mater which account it queries. Is this problematic for the random generation/usage?

File: fvm/environment/transaction_info.go [hunk 2]
```diff
@@ -19,6 +20,25 @@ type TransactionInfoParams struct {
...
+		Script: code,
+	}
+}
+
+func NewScriptInfoParamsWithCode(code []byte) *ScriptInfoParams {
+	return &ScriptInfoParams{
+		ID:     flow.HashToID(hash.DefaultComputeHash(code)),
+		Script: code,
+	}
+}
```
Comment (janezpodhostnik) [inline fvm/environment/transaction_info.go:40]: this file is named `transaction_info.go` maybe move this to its own file

File: fvm/script.go (hunk 1 above)
Comment (tarakby) [reply fvm/script.go:27]: Good point!
I don't know if some applications require randoms to be diversified per script argument. There may be some application with such requirement, so I added this: https://github.com/onflow/flow-go/pull/4648/commits/a00d5f0204d4c08afb886417b31576cb9a9d7b02

However I didn't update the other `ScriptProcedure` ID. I don't know where that is used.

File: fvm/environment/transaction_info.go (hunk 2 above)
Comment (tarakby) [reply fvm/environment/transaction_info.go]: https://github.com/onflow/flow-go/pull/4648/commits/3459ec75b47550deb6d63955374dcad458324029

**Review by janezpodhostnik — APPROVED** [review]

**Review by ramtinms — APPROVED** [review]
Looks good to me, just as a minor comment, I would suggest we pass in the data we need from the protocol state instead of injecting the github.com/onflow/flow-go/state/protocol into the query execution engine, given that long term this component would be run by archive nodes, they might not have access to full protocol state or might do it differently.

**Comment by tarakby** [comment]
@ramtinms that makes sense, let me see how I can refactor this.

**Comment by tarakby** [comment]
@ramtinms I added a minimal interface that is passed to `QueryExecutor` instead of all protocol state. Here is the change https://github.com/onflow/flow-go/pull/4648/commits/fbebac260a02ef085720df36446d968fe92db696. 
Let me know if that's what you wanted.

**Pull request merged**